### ISSUE TITLE
HTTP API path to restart go2rtc (and maybe ffmpeg)

### ISSUE THREAD

**Issue by esand** [blakeblackshear/frigate#15725]
**Describe what you are trying to accomplish and why in non technical terms**
I want to be able to restart go2rtc (and maybe ffmpeg - see additional context) processes that may have hung or are consuming high cpu/ram so that I can restore frigate health without restarting all of frigate.

**Describe the solution you'd like**
Since go2rtc and ffmpeg are separate processes, there are times where those programs may misbehave. I have noticed that go2rtc can end up consuming very high CPU, and it requires restarting all of Frigate for it to kill the process and restart it. It would be much nicer if I could have an API call that would just restart a specific process.

**Describe alternatives you've considered**
Currently the only API call that can be made is to restart Frigate, so that's the only solution available if go2rtc starts consuming high CPU (that or restarting the docker container).

**Additional context**
I'm not certain if ffmpeg has any need to be restarted, and since it can be per (sub)stream per camera, it might be more complex to add that. At the very least however, as Frigate spawns a go2rtc instance which is used for accessing all the cameras, it would be in the best interest to have ways to ensure its constant health by being able to restart it if needed without restarting all of Frigate.

**Comment by NickM-27** [comment]
Frigate already runs a health check on go2rtc, perhaps we can add more checks to this but that can cause false positive restarts. I'm not sure if an API to kill whatever process is really a good idea, it has its own security implications and in general is addressing a symptom instead of just fixing the actual issue which may just be an error in the config or some other problem in go2rtc

**Comment by esand** [comment]
Ok, that's quite interesting that it has its own health checks, because just now I checked my instance of Frigate and it was reporting 230% cpu usage, and when I checked the system details, go2rtc was consuming ~230% cpu constantly. I noticed this issue in previous releases and I believe someone did some work with regards to fixing hanging connections that caused runaway CPU usage, but it appears that there may still be a few edge cases.

Does the health check still get used if someone has a custom go2rtc binary (as I do in my case)? Perhaps that's the issue?

The reason for requesting the ability to restart go2rtc is, while I totally agree it isn't actually fixing the underlying issue, because go2rtc isn't maintained by Frigate. If it so happens that there is some issue in go2rtc and restarting it gets it back to a working state - even just for a while - I feel that's possibly about as much as you can do (other than telling the user to try a different version).

I just figured that it might be an easy thing to add - and only in the API so just for "advanced" users who know why they'd be restarting it.

**Comment by NickM-27** [comment]
> Does the health check still get used if someone has a custom go2rtc binary (as I do in my case)? Perhaps that's the issue?

Yes it works regardless. As I said it doesn't base things off of cpu usage because there are cases where high cpu usage is working as expected. 

We can see what other maintainers think

**Comment by hawkeye217** [comment]
I'm with @NickM-27, the benefits don't outweigh the risks with possible security implications and the fact that it's addressing a symptom rather than fixing the issue.

For "advanced" users, there are lots of ways to achieve this already without it built into Frigate - perhaps with some kind of bash script to kill go2rtc that's connected to a HA command line switch or something similar.

**Comment by esand** [comment]
Not to beat a dead horse, but are there security implications with the existing Frigate restart API call that exists? I believe you can even just publish frigate/restart via MQTT to kill the entire container. If we're talking DoS or something, this could be used to prevent recording.

If there's the possibility of having the process watchdogs look at average CPU/ram usage over time - that would be nice. When it does happen on my system, it impacts recording and viewing.

**Comment by hawkeye217** [comment]
Does a simple `pkill -f go2rtc` not solve your particular need? 

As Nick said, for some users, high CPU usage with go2rtc is normal, so that wouldn't be universal criteria for a watchdog.

**Comment by esand** [comment]
I'm sure it would, but I just haven't tried doing it yet since that would mean me logging in to a bunch of things to issue the command.

I'll close this FR since it seems everyone poopoo'd my idea :cry: and hope that whatever the issue is with go2rtc gets resolved soon (I think it has to do with 2-way audio streams and consumers).

**Comment by Trexano99** [comment]
I am facing the same exact problem as you and i have found that you can **restart go2rtc without restarting all Frigate**.
That is possible through a POST to go2rtc and for that i suggest you to create a `rest_command` in your `configuration.yaml` to be able to call the go2rtc restart as a service (and then also from an automation).
Be only aware to have enabled the go2rtc API port in Frigate (in my case the default 1984).

This is what you can place in your `config.yaml`:
```
rest_command:
  restart_go2rtc:
    method: POST
    url: "http://ccab4aaf-frigate:1984/api/restart"
```

Then you will be able to call the action from automations (or scripts) when the cpu load of frigate suddenly increase based on your needs:
```
action: rest_command.restart_go2rtc
```

**Comment by esand** [comment]
@Trexano99 Simply brilliant. I had read the API info for go2rtc but the main doc didn't mention any restart call. I just checked the yaml doc that defines the API calls and seems the doc misses a lot of things.

I just set up a RESTful command pretty much like you mentioned and tested it out - works great (so long as go2rtc is responsive). I've got it in an automation so if it ever happens again, I'll get to test it in a real world situation.

BTW, if you're running Frigate in a docker container and it has a local network IP that Home Assistant can reach, you don't need to expose port 1984. Exposing ports in a docker container is only to map them to the docker system's main IP, which in my case I didn't need to do.
